tango_simlib/utilities/validate_device.py

```diff
@@ -0,0 +1,415 @@
+    spec_data_names = {i["name"] for i in spec_data}
+    dev_data_names = {i["name"] for i in dev_data}
+    if spec_data_names != dev_data_names:
+        diff = spec_data_names.difference(dev_data_names)
+        if diff:
+            diff = sorted(diff)
+            issues.append(
+                "{} differs, [{}] specified but missing in device".format(
+                    type_str, ",".join(diff)
+                )
+            )
+
+        if bidirectional:
+            diff = dev_data_names.difference(spec_data_names)
+            if diff:
+                diff = sorted(diff)
+                issues.append(
+                    "{} differs, [{}] present in device but not specified".format(
+                        type_str, ",".join(diff)
+                    )
+                )
+
+    # Check that the commands/attributes (by name) that are in both the spec and device
+    # are the same
+    mutual_names = spec_data_names.intersection(dev_data_names)
+    mutual_spec_data = filter(lambda x: x["name"] in mutual_names, spec_data)
+    mutual_dev_data = filter(lambda x: x["name"] in mutual_names, dev_data)
+    for spec, dev in zip(mutual_spec_data, mutual_dev_data):
+        issues.extend(check_single_dict_differences(spec, dev, type_str, bidirectional))
+
+    return issues
+
+
+def check_single_dict_differences(spec, dev, type_str, bidirectional):
+    """Compare a single attribute/property
+
+    Parameters
+    ----------
+    spec : dict
+        A single Attribute/Command dictionary form the specficication
+        E.g {'disp_level': 'OPERATOR',
+             'doc_in': 'ON/OFF',
+             'doc_out': 'Uninitialised',
+             'dtype_in': 'DevBoolean',
+             'dtype_out': 'DevVoid',
+             'name': 'Capture'}
+
+    dev : dict
+        A single Attribute/Command dictionary form the device
+        E.g {'disp_level': 'OPERATOR',
+             'doc_in': 'ON/OFF',
+             'doc_out': 'Uninitialised',
+             'dtype_in': 'DevBoolean',
+             'dtype_out': 'DevVoid',
+             'name': 'Capture'}
+
+    type_str : str
+        Either "Command" or "Attribute"
+
+    bidirectional: bool
+        Whether to include details on the device that is not in the specification
+
+    Returns
+    -------
+    list
+        A list of strings describing the issues, empty list for no issues
+    """
+    assert spec["name"] == dev["name"]
+    issues = []
+
+    if spec != dev:
+        spec_keys = set(spec.keys())
+        dev_keys = set(dev.keys())
+
+        keys_not_in_spec = spec_keys.difference(dev_keys)
+        keys_not_in_dev = dev_keys.difference(spec_keys)
+        mutual_keys = spec_keys.intersection(dev_keys)
+
+        if keys_not_in_spec:
+            keys_not_in_spec = sorted(keys_not_in_spec)
+            issues.append(
+                "{} [{}] differs, specification has keys [{}] but it's "
+                "not in device".format(type_str, spec["name"], ",".join(keys_not_in_spec))
```

Comment:
Please fix the line length.

Reply:
Too long ? The linter in the pipeline does a check for lines that are too long.

See an old failure here: http://ci.camlab.kat.ac.za/job/tango_simlib-multibranch/job/user%252Fjohan%252FSAR-113_add_tango_device_spec_validation/2/console

Reply:
My bad. It looked a bit longer on my editor.